File: python/questions/hIndex.py
from typing import List


class Solution:
    def hIndex(self, citations: List[int]) -> int:
        lc = len(citations)
        rc = [0] * (lc + 1)
        for c in citations:
            if c >= lc:
                rc[lc] += 1
            else:
                rc[c] += 1
        rs = 0
        for p in range(lc, -1, -1):
            rs += rc[p]
            if rs >= p:
                return p
        return 0

File: cpp/questions/hIndex.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int hIndex(vector<int>& citations) {
        int cs = citations.size();
        std::vector<int> rc(cs+1, 0);
        for (const auto& c: citations) {
            if (c >= cs) {
                rc[cs] += 1;
            } else {
                rc[c] += 1;
            }
        }
        int rs = 0;
        for (int p=cs; p>=0; --p) {
            rs += rc[p];
            if (rs >= p) {
                return p;
            }
        }
        return 0;
    }
};
